From 23471222f5332d17905aa46c28cab98ec4ea1988 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Fri, 28 Jul 2017 10:30:35 +0800
Subject: [PATCH 2260/5242] MLK-16093-3 cpufreq: imx8mq: add cooling device
 support

commit  3ec464d45f44a89e4e6010604854e2306306f7ff from
https://source.codeaurora.org/external/imx/linux-imx.git

Add i.MX8MQ cooling device support, when temperature
exceeds passive threshold, cpu-freq will drop to lowest
set-point, and once temperature drops below passive
threshold, cpu-freq will restore.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/cpufreq/imx8mq-cpufreq.c |   22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/drivers/cpufreq/imx8mq-cpufreq.c b/drivers/cpufreq/imx8mq-cpufreq.c
index 1765b47..0457ee2 100644
--- a/drivers/cpufreq/imx8mq-cpufreq.c
+++ b/drivers/cpufreq/imx8mq-cpufreq.c
@@ -9,6 +9,7 @@
 #include <linux/clk.h>
 #include <linux/cpu.h>
 #include <linux/cpufreq.h>
+#include <linux/cpu_cooling.h>
 #include <linux/err.h>
 #include <linux/module.h>
 #include <linux/slab.h>
@@ -28,6 +29,7 @@
 static struct clk *arm_pll_clk;
 static struct clk *arm_pll_out_clk;
 static struct clk *sys1_pll_800m_clk;
+struct thermal_cooling_device *cdev;
 
 static int imx8mq_set_target(struct cpufreq_policy *policy, unsigned int index)
 {
@@ -68,6 +70,24 @@ static int imx8mq_set_target(struct cpufreq_policy *policy, unsigned int index)
 	return ret;
 }
 
+static void imx8mq_cpufreq_ready(struct cpufreq_policy *policy)
+{
+	struct device_node *np = of_get_cpu_node(policy->cpu, NULL);
+
+	if (of_find_property(np, "#cooling-cells", NULL)) {
+		cdev = of_cpufreq_cooling_register(np, policy);
+
+		if (IS_ERR(cdev) && PTR_ERR(cdev) != -ENOSYS) {
+			pr_err("cpu%d is not running as cooling device: %ld\n",
+					policy->cpu, PTR_ERR(cdev));
+
+			cdev = NULL;
+		}
+	}
+
+	of_node_put(np);
+}
+
 static int imx8mq_cpufreq_init(struct cpufreq_policy *policy)
 {
 	int ret;
@@ -92,6 +112,7 @@ static int imx8mq_cpufreq_init(struct cpufreq_policy *policy)
 	.get = cpufreq_generic_get,
 	.init = imx8mq_cpufreq_init,
 	.name = "imx8mq-cpufreq",
+	.ready = imx8mq_cpufreq_ready,
 	.attr = cpufreq_generic_attr,
 #ifdef CONFIG_PM
 	.suspend = cpufreq_generic_suspend,
@@ -199,6 +220,7 @@ static int imx8mq_cpufreq_probe(struct platform_device *pdev)
 
 static int imx8mq_cpufreq_remove(struct platform_device *pdev)
 {
+	cpufreq_cooling_unregister(cdev);
 	cpufreq_unregister_driver(&imx8mq_cpufreq_driver);
 	dev_pm_opp_free_cpufreq_table(cpu_dev, &freq_table);
 	if (free_opp)
-- 
1.7.9.5

